File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> list;
    int sum;
    vector<vector<int>> combinationSum3(int k, int n) {
        backtracking(1, n, k);
        return res;
    }

    void backtracking(int start, int n, int k) {
        if (list.size() == k) {
            if (sum == n) {
                vector<int> copy = list;
                res.push_back(copy);
            }
            
            return;
        }

        for (int i = start; i <= 9; i++) {
            if (sum + i > n) {
                continue;
            }

            list.push_back(i);
            sum += i;
            backtracking(i + 1, n, k);
            sum -= i;
            list.pop_back();
        }
    }
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def __init__(self):
        self.path = []
        self.res = []

    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        self.backtracking(1, 0, k, n)
        return self.res

    def backtracking(self, curr, currSum, k, n):
        if len(self.path) == k:
            if currSum == n:
                self.res.append(copy.copy(self.path))

            return

        for i in range(curr, 10):
            if currSum + i > n:
                return
            
            self.path.append(i)
            self.backtracking(i + 1, currSum + i, k, n)
            self.path.pop()


            